server/app/views/components/FieldWithLabel.java

```diff
@@ -246,107 +245,166 @@ public FieldWithLabel showFieldErrors(boolean showFieldErrors) {
-      // Setting inputmode to decimal gives iOS users a more accessible keyboard
-      fieldTag.attr("inputmode", "decimal");
-
-      // Setting step to any disables the built-in HTML validation so we can use our
-      // custom validation message to enforce integers.
-      fieldTag.attr("step", "any");
-
-      // Set min and max values for client-side validation
-      if (this.minValue.isPresent()) {
-        fieldTag.attr("min", minValue.getAsLong());
-      }
-      if (this.maxValue.isPresent()) {
-        fieldTag.attr("max", maxValue.getAsLong());
-      }
+  }
 
-      // For number types, only set the value if it's present since there is no empty string
-      // equivalent for numbers.
-      if (this.fieldValueNumber.isPresent()) {
-        fieldTag.withValue(String.valueOf(this.fieldValueNumber.getAsLong()));
-      }
-    } else {
-      fieldTag.withValue(this.fieldValue);
+  private LabelTag genLabelTag() {
+
+    return label()
+        .withFor(this.id)
+        // If the text is screen-reader text, then we want the label to be screen-reader
+        // only.
+        .withClass(labelText.isEmpty() ? Styles.SR_ONLY : BaseStyles.INPUT_LABEL)
+        .withText(labelText.isEmpty() ? screenReaderText : labelText);
+  }
+
+  private DivTag buildBaseContainer(Tag fieldTag, Tag labelTag, String fieldErrorsId) {
+    return div(
+            labelTag,
+            div(fieldTag, buildFieldErrorsTag(fieldErrorsId))
+                .withClasses(Styles.FLEX, Styles.FLEX_COL))
+        .withClasses(
+            StyleUtils.joinStyles(referenceClassesBuilder.build().toArray(new String[0])),
+            BaseStyles.FORM_FIELD_MARGIN_BOTTOM);
+  }
+
+  private boolean hasFieldErrors() {
+    return !fieldErrors.isEmpty() && showFieldErrors;
+  }
+
+  protected void numberTagApplyAttrs(InputTag fieldTag) {
+    // Setting inputmode to decimal gives iOS users a more accessible keyboard
+    fieldTag.attr("inputmode", "decimal");
+
+    // Setting step to any disables the built-in HTML validation so we can use our
+    // custom validation message to enforce integers.
+    fieldTag.withStep("any");
+
+    // Set min and max values for client-side validation
+    if (this.minValue.isPresent()) {
+      fieldTag.attr("min", minValue.getAsLong());
+    }
+    if (this.maxValue.isPresent()) {
+      fieldTag.attr("max", maxValue.getAsLong());
     }
 
-    boolean hasFieldErrors = !fieldErrors.isEmpty() && showFieldErrors;
-    if (hasFieldErrors) {
-      fieldTag.attr("aria-invalid", "true");
-      fieldTag.attr("aria-describedBy", fieldErrorsId);
+    // For number types, only set the value if it's present since there is no empty string
+    // equivalent for numbers.
+    if (this.fieldValueNumber.isPresent()) {
+      fieldTag.withValue(String.valueOf(this.fieldValueNumber.getAsLong()));
     }
+  }
 
+  private <T extends Tag<T> & IName<T> & IDisabled<T>> void generalApplyAttrsClassesToTag(
+      T fieldTag, boolean hasFieldErrors) {
```

Comment:
Given that this is a private method that knows the details on how the relevant FieldWithLabels are constructed, I'd be ok losing a bit of type safety here in order to avoid having long generic type declarations.

Using a `.attr` or `.condAttr` call here with a comment as to why we're not using a more type-safe version seems like it'd be an ok tradeoff. This would reduce the number of potentially moving parts in this CL.

RE separate subclasses per field type: Maybe. I'd keep it separate from this CL to avoid adding extra complexity.